typing/ctype.ml

```diff
@@ -1845,30 +1845,34 @@ let is_contractive env p =
 
 exception Occur
 
-let rec occur_rec env allow_recursive visited ty0 ty =
-  if eq_type ty ty0 then raise Occur;
-  match get_desc ty with
-    Tconstr(p, _tl, _abbrev) ->
-      if allow_recursive && is_contractive env p then () else
-      begin try
-        if TypeSet.mem ty visited then raise Occur;
-        let visited = TypeSet.add ty visited in
-        iter_type_expr (occur_rec env allow_recursive visited ty0) ty
-      with Occur -> try
-        let ty' = try_expand_head try_expand_safe env ty in
-        (* This call used to be inlined, but there seems no reason for it.
-           Message was referring to change in rev. 1.58 of the CVS repo. *)
-        occur_rec env allow_recursive visited ty0 ty'
-      with Cannot_expand ->
-        raise Occur
-      end
-  | Tobject _ | Tvariant _ ->
-      ()
-  | _ ->
-      if allow_recursive ||  TypeSet.mem ty visited then () else begin
-        let visited = TypeSet.add ty visited in
-        iter_type_expr (occur_rec env allow_recursive visited ty0) ty
-      end
+let rec occur_rec env mark allow_recursive visited ty0 ty =
+  if not_marked_node mark ty then begin
+    if eq_type ty ty0 then raise Occur;
+    begin match get_desc ty with
+      Tconstr(p, _tl, _abbrev) ->
+        if allow_recursive && is_contractive env p then () else
+        begin try
+          if TypeSet.mem ty visited then raise Occur;
+          let visited = TypeSet.add ty visited in
+          iter_type_expr (occur_rec env mark allow_recursive visited ty0) ty
+        with Occur -> try
+          let ty' = try_expand_head try_expand_safe env ty in
+          (* This call used to be inlined, but there seems no reason for it.
+            Message was referring to change in rev. 1.58 of the CVS repo. *)
+          occur_rec env mark allow_recursive visited ty0 ty'
+        with Cannot_expand ->
+          raise Occur
+        end
+    | Tobject _ | Tvariant _ ->
+        ()
+    | _ ->
+        if allow_recursive ||  TypeSet.mem ty visited then () else begin
```

Comment:
I don't know where to leave this comment so I'll put it here :

I'm surprised by this line. Here we say that if we see a recursive type when checking that we don't create one we can just ignore it. I know that rejecting it at that moment without having a good location for an error message might be problematic however isn't it better to reject it at that moment than to accept the code ?

The only tests in the testsuite affected by adding an `assert allow_recursive` here are tests identified as bugs

Reply:
The test is `typing-polyvariants-bug-2/pr3918.ml` is not exactly a bug but it is probably fine to reject a recursive type which is not allowed due to a missing cmi file.

Reply:
@garrigue , can you think of any other circumstances when we may have a cycle not guarded by an object or variant types?
Otherwise, I would be tempted to consider that it is better to fail on those kind of unguarded cycles created by missing `cmi`s.

Reply:
I wrote some documentation on what I knew of recursive type situations in OCaml in typedecl.ml (to document `check_well_founded`):

https://github.com/ocaml/ocaml/blob/8f7c58f8195abfcc818bd2ff031acad5a67fa244/typing/typedecl.ml#L680

The documentation agrees with your intuition that in absence of `-rectypes`, recursive types are only allowed to go through polymorphic variants and object types. (To check this, look for the occurrences of " guarded" in the documentation, which list the only sources of allowed cycles; when `-rectypes` is used, the "rectypes-guarded" constructs are also allowed.)

Reply:
For whether it is ok to fail when discovering a pre-existing cycle, this requires checking a number of things: namely, interactions with the counter-example mode (which is allowed to create such cycles), and mixing of files compiled with an without `-rectypes`. I don't have all the details in mind, but this would depend on this.
Also, logically, it should probably be an assertion rather than an occur check error, since it would not be supposed to happen.